File: src/my_py_pkg/my_py_pkg/minimal_subscriber.py
import rclpy
from rclpy.executors import ExternalShutdownException
from example_interfaces.msg import String
from rclpy.node import Node


class MinimalSubscriber(Node):
    def __init__(self):
        super().__init__("minimal_subscriber")
        self.subscription = self.create_subscription(
            String, "my_topic", self._listener_callback, 10
        )
        self.subscription  # prevent unused variable warning

    def _listener_callback(self, msg):
        self.get_logger().info('I heard: "%s"' % msg.data)


def main(args=None):
    rclpy.init(args=args)
    minimal_subscriber = MinimalSubscriber()
    try:
        rclpy.spin(minimal_subscriber)
    except KeyboardInterrupt:
        pass
    except ExternalShutdownException:
        pass
    finally:
        minimal_subscriber.destroy_node()
        rclpy.shutdown()


if __name__ == "__main__":
    main()

File: src/my_cpp_pkg/src/minimal_subscriber.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/msg/string.hpp"


/**
 * Subscriber example that prints message to console
 */
class MinimalSubscriber : public rclcpp::Node
{
public:

/**
 * Constructor (call Node class with name)
 */
    MinimalSubscriber()
    : Node("minimal_subscriber")
    {
        // create subscription
        subscription_ = this->create_subscription<example_interfaces::msg::String>(
            "my_topic", 10, std::bind(
                &MinimalSubscriber::listener_callback,
                this,
                std::placeholders::_1));
    }

private:
    void listener_callback(const example_interfaces::msg::String & msg) const
    {
        RCLCPP_INFO(this->get_logger(), "I heard: '%s'", msg.data.c_str());
    }
    rclcpp::Subscription<example_interfaces::msg::String>::SharedPtr subscription_;
};

/**
 * Main function
 */
int main(int argc, char * argv[])
{
    /** Initialize ROS */
    rclcpp::init(argc, argv);

    // Create a shared pointer to the MinimalSubscriber node
    auto node = std::make_shared<MinimalSubscriber>();

    // Spin the node to process incoming messages
    rclcpp::spin(node);

    // Shutdown ROS
    rclcpp::shutdown();
    return 0;
}
